bundles/org.openhab.binding.epsonprojector/src/main/java/org/openhab/binding/epsonprojector/internal/handler/EpsonProjectorHandler.java

```diff
@@ -0,0 +1,417 @@
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.epsonprojector.internal.EpsonProjectorCommandException;
+import org.openhab.binding.epsonprojector.internal.EpsonProjectorCommandType;
+import org.openhab.binding.epsonprojector.internal.EpsonProjectorDevice;
+import org.openhab.binding.epsonprojector.internal.EpsonProjectorException;
+import org.openhab.binding.epsonprojector.internal.configuration.EpsonProjectorConfiguration;
+import org.openhab.binding.epsonprojector.internal.enums.AspectRatio;
+import org.openhab.binding.epsonprojector.internal.enums.Background;
+import org.openhab.binding.epsonprojector.internal.enums.ColorMode;
+import org.openhab.binding.epsonprojector.internal.enums.Gamma;
+import org.openhab.binding.epsonprojector.internal.enums.Luminance;
+import org.openhab.binding.epsonprojector.internal.enums.PowerStatus;
+import org.openhab.binding.epsonprojector.internal.enums.Switch;
+import org.openhab.core.io.transport.serial.SerialPortManager;
+import org.openhab.core.library.types.DecimalType;
+import org.openhab.core.library.types.OnOffType;
+import org.openhab.core.library.types.StringType;
+import org.openhab.core.thing.Channel;
+import org.openhab.core.thing.ChannelUID;
+import org.openhab.core.thing.Thing;
+import org.openhab.core.thing.ThingStatus;
+import org.openhab.core.thing.ThingStatusDetail;
+import org.openhab.core.thing.ThingTypeUID;
+import org.openhab.core.thing.binding.BaseThingHandler;
+import org.openhab.core.types.Command;
+import org.openhab.core.types.RefreshType;
+import org.openhab.core.types.State;
+import org.openhab.core.types.UnDefType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link EpsonProjectorHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Pauli Anttila, Yannick Schaus - Initial contribution
+ * @author Michael Lobstein - Improvements for OH3
+ */
+@NonNullByDefault
+public class EpsonProjectorHandler extends BaseThingHandler {
+    private static final int DEFAULT_POLLING_INTERVAL_SEC = 10;
+
+    private final Logger logger = LoggerFactory.getLogger(EpsonProjectorHandler.class);
+
+    private @Nullable ScheduledFuture<?> pollingJob;
+    private final EpsonProjectorDevice device;
+
+    private boolean isPowerOn = false;
+    private int pollingInterval = DEFAULT_POLLING_INTERVAL_SEC;
+
+    public EpsonProjectorHandler(Thing thing, SerialPortManager serialPortManager) {
+        super(thing);
+
+        ThingTypeUID thingTypeUID = thing.getThingTypeUID();
+        if (THING_TYPE_PROJECTOR_SERIAL.equals(thingTypeUID)) {
+            device = new EpsonProjectorDevice(serialPortManager, getConfigAs(EpsonProjectorConfiguration.class));
+        } else {
+            device = new EpsonProjectorDevice(getConfigAs(EpsonProjectorConfiguration.class));
+        }
```

Comment:
You should move this to initialize() method, otherwise user configuration changes are not taken in use.

Reply:
Hmm, but then device still needs to be set to something here to satisfy the notnullbydefault. And EpsonProjectorDevice needs a connector in the constructor which is why I had done the EpsonProjectorDefaultConnector.

Reply:
Well, notnullbydefault just forces you to think which field can be null and which not. In this case device can't be initialised as it depends on the thing configuration an it can change during thing life time.

So it should be fine to be nullable

private @Nullable EpsonProjectorDevice device;

or if you want to avoid nulls, you could make it optional

private Optional<EpsonProjectorDevice> device = Optional.empty();

Reply:
OK, fixed (using Optional).